Check string comparison size

mmisono opened the pull request:
In the debug build, the following assertion failure occurs:

```
% sudo ./src/bpftrace -e 'BEGIN /comm == "aaaaaaaaaaaaaaaaaaaaaaaaa"/ {}'
bpftrace: /home/ubuntu/work/bpftrace/bpftrace/src/ast/irbuilderbpf.cpp:683: llvm::Value *bpftrace::ast::IRBuilderBPF::CreateStrncmp(llvm::Value *, llvm::Value *, bpftrace::AddrSpace, std::string, uint64_t, const bpftrace::location &, bool): Assertion `valp->getElementType()->isArrayTy() && valp->getElementType()->getArrayNumElements() >= n && valp->getElementType()->getArrayElementType() == getInt8Ty()' failed.
```

The problem part is https://github.com/iovisor/bpftrace/blob/a03b5c4bbb12a25345a2fdec78374963ff936131/src/ast/irbuilderbpf.cpp#L682
and the reason is that the the size of "comm" is 16, but the size of
string literal is bigger than that.

To fix the problem, check the string size in the semantic analyzer and
error if the size of string literal exeeds the that of the other. Now
the above exmaple becomes

```
% sudo ./src/bpftrace -e 'BEGIN /comm == "aaaaaaaaaaaaaaaaaaaaaaaaa"/ {}'
stdin:1:7-43: ERROR: The right string literal has the longer length of the size of the left (16 < 25)
BEGIN /comm == "aaaaaaaaaaaaaaaaaaaaaaaaa"/ {}
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

<!--
Please provide a description of your change below this comment.

Then please complete the checklist.
-->

##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests


---------------

I wonder if this could cause a regression. I don't think so, but...

fbs commented on src/ast/semantic_analyser.cpp:
```diff
@@ -1499,6 +1499,32 @@ void SemanticAnalyser::visit(Binop &binop)
           << " operator can not be used on expressions of types " << lhs << ", "
           << rhs;
     }
+    else if (binop.op == Parser::token::EQ && !binop.left->is_literal &&
```
I feel like emitting a warning and always return false should be better here. Don't want to block a valid script.

mmisono replied:
sure, will fix.

fbs commented on src/ast/semantic_analyser.cpp:
```diff
@@ -1499,6 +1499,32 @@ void SemanticAnalyser::visit(Binop &binop)
+      auto rlen = bpftrace_.get_string_literal(binop.right).size();
+      if (rlen > llen)
+      {
+        LOG(ERROR, binop.left->loc + binop.loc + binop.right->loc, err_)
```
Dont think we have to keep two cases here. Something simple like this should be enough:

```
Literal is longer than the variable string, condition will always be false
```

mmisono commented:
Updated:
- Changed the semantic analyzer so that it emits a warning instead of making an error. Also, make the check simple.
- In the CreateStrcmp, skip comparison if the condition is always false. Also, fix assertion so that it does not assert that case.

fbs reviewed (approved):
Looks good.

Having a warning test might be useful tho, just to catch this case in a test somewhere.

mmisono commented:
Added test for string comparisons
